### DISCUSSION TITLE
ERROR: Could not resolve "path" when trying to use preval.macro in edge middleware

### DISCUSSION THREAD

**Discussion by Jack-Sibley** [vercel/vercel#8952] [Help] [answered]
I have an astro project using pnpm, and I'm trying to write edge middleware that handles some redirect functionality. Whenever I import and use any kind of [babel-plugin-macros](https://github.com/kentcdodds/babel-plugin-macros) compatible macro. I get the following wall of errors.

```
vc-file-system:node_modules/.pnpm/babel-plugin-macros@3.1.0/node_modules/babel-plugin-macros/dist/index.js:3:18: ERROR: Could not resolve "path"
vc-file-system:node_modules/.pnpm/babel-plugin-preval@5.1.0/node_modules/babel-plugin-preval/dist/helpers.js:11:43: ERROR: Could not resolve "path"
vc-file-system:node_modules/.pnpm/cosmiconfig@7.1.0/node_modules/cosmiconfig/dist/Explorer.js:8:43: ERROR: Could not resolve "path"
vc-file-system:node_modules/.pnpm/cosmiconfig@7.1.0/node_modules/cosmiconfig/dist/ExplorerBase.js:9:43: ERROR: Could not resolve "path"
vc-file-system:node_modules/.pnpm/cosmiconfig@7.1.0/node_modules/cosmiconfig/dist/ExplorerSync.js:8:43: ERROR: Could not resolve "path"
```

Anyone had anything like this before or know of a solution?

**Comment by Jack-Sibley** [comment, marked as answer]
Update: OK this is mostly a preval.macro issue. It leaves it's own import in the output file which of course imports a bunch of node modules that aren't available on the edge. My hacky solution was just to add a `sed` command to my build script to delete the top line of my `middleware.js` (which contained the macro import).

I'm going to look into PRing a fix to preval as I think this would a very useful tool when working on the edge.

**Reply by amyegan** [reply]
Thanks for sharing this, Jack!
